### PR TITLE
Feat/add on click

### PR DESCRIPTION
**Pull request by Solnick**

### CODE REVIEW

File: src/Board.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    const boardContainer = document.querySelector('#board-container');
+    for (let i = 0; i < this.height; i++) {
+      this.tiles[i] = [];
+    }
+    for (let i = 0; i < this.height; i++) {
+      const row = document.createElement('div');
+      row.classList.add(`row`);
+      boardContainer.append(row);
+      for (let j = 0; j < this.width; j++) {
+        this.tiles[j][i] = new Tile(j, i, this);
```
Comment (Solnick) [inline src/Board.js:29]: in this function you are creating new tiles. It does not like it is function for rendering the board. so maybe let's rename the `render` function to `initializeTiles`

File: src/Board.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        row.append(this.tiles[j][i].tileContainer);
+      }
+    }
+  };
+  createInitialCharacters = () => {
+    const numberOfCharacters = Math.round(this.width * this.height * 0.3);
+    for (let i = 0; i < numberOfCharacters; i++) {
+      let RandomOrganismClass =
+        classesList[Math.floor(Math.random() * classesList.length)];
```
Comment (Solnick) [inline src/Board.js]:
```suggestion
const randomClassIndex = Math.floor(Math.random() * classesList.length)
...
        classesList[randomClassIndex];
```

File: src/Board.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    return foundTile;
+  };
+  round = async () => {
+    const organismsOnBoard = this.getOrganisms();
+    // console.log(organismsOnBoard);
+    this.sortedOrganismsOnBoard = this.sortOrganisms(organismsOnBoard);
+    console.log(this.sortedOrganismsOnBoard);
+    for (let i = 0; i < this.sortedOrganismsOnBoard.length; i++) {
+      // if (this.sortedOrganismsOnBoard[i].currentOrganism.isAlive === true){
```
Comment (Solnick) [inline src/Board.js]: Don't we need this if statement?

File: src/Board.js (hunk 1 above)
Comment (nattsnb) [reply src/Board.js:29]: done

File: src/Board.js (hunk 3 above)
Comment (nattsnb) [reply src/Board.js]: nope

File: src/Board.js (hunk 2 above)
Comment (nattsnb) [reply src/Board.js]: changed. as well in eventlistner for adding random organism on click.

File: src/Board.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    // move left.down
+    if (x > minimalStep && y < yMax) {
+      arrayOfNearestTiles.push(
+        this.tiles[x - organism.numberOfSteps][y + organism.numberOfSteps],
+      );
+    }
+    // move left.straight
+    if (x > minimalStep) {
+      arrayOfNearestTiles.push(this.tiles[x - organism.numberOfSteps][y]);
+    }
```
Comment (Solnick) [inline src/Board.js]: all those if statements are code duplicates. Create a generic function to do it for you. Just remember to return a new array every time rather than modifying arguments (changing arguments of array/object type is a bad practice - it makes the code hard to follow when reading)

File: src/Board.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      await this.sortedOrganismsOnBoard[i].action();
+    }
+    const organismsAfterRound = this.getOrganisms();
+    return organismsAfterRound.find(Board.isOrganismAPlayer);
+  };
+  getOrganisms = () => {
+    let organismsOnBoard = [];
+    for (let y = 0; y < this.tiles.length; y++) {
+      for (let x = 0; x < this.tiles.length; x++) {
+        if (this.tiles[x][y].currentOrganism) {
```
Comment (Solnick) [inline src/Board.js]: Maybe if we would query the tile
```
tile.getCurrentOrganism()
```
We would be able to omit the if statement outside of the tile
```
getCurrentOrganism() {
  return this.currentOrganism.isAlive ?  this.currentOrganism : null
}
```

File: src/index.js [hunk 6]
```diff
@@ -1,3 +1,19 @@
...
+
+// console.log(board.tiles);
+// console.log(board.tiles[1][2]
+
+let game = 0;
+const delay = (ms) => new Promise((res) => setTimeout(res, ms));
+
+let isPlayerAlive = true;
+
+while (isPlayerAlive) {
```
Comment (Solnick) [inline src/index.js]:
```suggestion

while (board.isPlayerAlive()) {
```

File: src/player/Player.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    if (evt.code === 'KeyD') {
+      console.log('KeyD');
+      const newX = this.activeTile.x + 1;
+      if (
+        this.tilesForAction.includes(this.board.tiles[newX][this.activeTile.y])
+      ) {
+        const coordinates = { x: newX, y: this.activeTile.y };
+        this.moveActiveTile(coordinates);
+      }
+    }
```
Comment (Solnick) [inline src/player/Player.js]: Duplicate code, we can wrap it in a function and call it 4 times

File: src/antelope/Antelope.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    super(board, startParameters);
+    this.createElement();
+  }
+
+  fight = (opponent) => {
+    // console.log(`I will fight`);
+    // animal looses
+    if (opponent.strength > this.strength) {
+      let odds = Math.random();
+      if (odds < 0.5) {
```
Comment (Solnick) [inline src/antelope/Antelope.js]:
```suggestion
      if (odds < this.chanceForFleeing) {
```

File: src/turtle/Turtle.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  };
+
+  constructor(board, startParameters) {
+    super(board, startParameters);
+    this.createElement();
+  }
+
+  action = () => {
+    let odds = Math.random();
+    if (odds < 0.25 && this.isAlive) {
```
Comment (Solnick) [inline src/turtle/Turtle.js]:
```suggestion
    if (odds < this.chancesOfMoving && this.isAlive) {
```

File: src/Animal.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+      return opponent instanceof this.constructor;
+    }
+  };
+  mate = (opponent) => {
+    const surroundingEmptyTiles = this.board.findEmptyTilesToMate(
+      this,
+      opponent,
+    );
+    // console.log(`I will mate`);
+    if (surroundingEmptyTiles.length > 0) {
```
Comment (Solnick) [inline src/Animal.js:63]: This should be a part of `shouldMate` method not the `mate` method itself

**Comment by Solnick** [comment]
image
@nattsnb The only lacking part is the modal which should pop up after the tile is clicked. You should be able to choose which organism to put on the tile rather than let the system decide randomly. 

There was a modal/dialog window screenshot in the task description. It might have been confusing, so I edited the instruction so it is clear that the dialog is required.

**Comment by Solnick** [comment]
After all the above changes I will do the review again. Please remove all the `console.logs` from the code and make sure that all the files are as clear as possible.

File: src/Animal.js (hunk 10 above)
Comment (nattsnb) [reply src/Animal.js:63]: hmm.. started to move the if statement to shouldMove(), but it seems to complicate the code even more. While I'm checking if there's a space for the child, I simultaneously pick the tile and use it later in mate(). If I move this to shouldMate(), I'll need to save picked tile somewhere to pass it to another function or look for the tile again. Any hints how to make it work?

File: src/turtle/Turtle.js (hunk 9 above)
Comment (nattsnb) [reply src/turtle/Turtle.js]: done

File: src/antelope/Antelope.js (hunk 8 above)
Comment (nattsnb) [reply src/antelope/Antelope.js]: done

File: src/index.js (hunk 6 above)
Comment (nattsnb) [reply src/index.js]: done

File: src/Board.js (hunk 5 above)
Comment (nattsnb) [reply src/Board.js]: done

does this question mark mean anything? not sure what is that piece of code.

File: src/player/Player.js (hunk 7 above)
Comment (nattsnb) [reply src/player/Player.js]: looks better now. tried to shorten it as much as possible

File: src/Board.js (hunk 4 above)
Comment (nattsnb) [reply src/Board.js]: I think I need help with this one. I spent ages trying to make it work, and really don't know how to make it better. 

The only part that duplicates is ```arrayOfNearestTiles.push(this.tiles``` this much of code. Does it really need separate function? 

So while I'm pushing into the array I am changing it, I guess. But how can I do it other way? I have an empty array and need to "iterate" around central tile to get all of them that exist around. So after checking if one tile exists I'm adding it to the array...

Comment (Solnick) [reply src/Board.js]:
```suggestion
function getTileBasedOnThePosition(shouldTheTileBeAdded, tile) {
  if (shouldTheTileBeAdded) {
    return [...arryOfNearestTiles, tile]
  }
}
let arrayOfNearestTiles = [];
arrayOfNearestTiles = getTileBasedOnThePosition(x > minimalStep && y > minimalStep, this.tiles[x - organism.numberOfSteps][y - organism.numberOfSteps])
//move straight.up
arrayOfNearestTiles = getTileBasedOnThePosition(y > minimalStep, this.tiles[x][y - organism.numberOfSteps])
// move right.up
arrayOfNearestTiles = getTileBasedOnThePosition(x < xMax && y > minimalStep, this.tiles[x + organism.numberOfSteps][y - organism.numberOfSteps])
// move right.straight
arrayOfNearestTiles = getTileBasedOnThePosition(x < xMax, this.tiles[x + organism.numberOfSteps][y])
// move right.down
arrayOfNearestTiles = getTileBasedOnThePosition(x < xMax && y < yMax, this.tiles[x + organism.numberOfSteps][y + organism.numberOfSteps])
//move straight.down
arrayOfNearestTiles = getTileBasedOnThePosition(y < yMax, this.tiles[x][y + organism.numberOfSteps])
// move left.down
arrayOfNearestTiles = getTileBasedOnThePosition(x > minimalStep && y < yMax, this.tiles[x - organism.numberOfSteps][y + organism.numberOfSteps])
// move left.straight
arrayOfNearestTiles = getTileBasedOnThePosition(x > minimalStep, this.tiles[x - organism.numberOfSteps][y])
```

File: src/Board.js (hunk 5 above)
Comment (Solnick) [reply src/Board.js]: google "conditional expression JS"

File: src/Animal.js (hunk 10 above)
Comment (Solnick) [reply src/Animal.js:63]: Alright let's leave it as is then ;)

File: src/Board.js [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    for (let y = 0; y < this.tiles.length; y++) {
+      for (let x = 0; x < this.tiles.length; x++) {
+        if (this.tiles[x][y].getCurrentOrganism()) {
+          organismsOnBoard.push(this.tiles[x][y].currentOrganism);
+        }
+      }
+    }
+    return organismsOnBoard;
+  };
+  sortOrganisms(organismsOnBoard) {
```
Comment (Solnick) [inline src/Board.js:83]: this should be an arrow function to keep consistency

File: src/Modal.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    const elementID = event.target.id;
+    if (organismIdToClassMapping[elementID]) {
+      this.addOrganismOnClick(organismIdToClassMapping[elementID]);
+    }
+    if (elementID === 'close-button') {
+      const modal = document.getElementById('addOrganismModal');
+      modal.style.display = 'none';
+    }
+  };
+  addOrganismOnClick = (whatClass) => {
```
Comment (Solnick) [inline src/Modal.js:65]:
```suggestion
  addOrganismOnClick = (Class) => {
```
`Class` is sufficient. We can Capitalize the argument since it is a class.

File: src/Modal.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+      icon.addEventListener('click', this.eventToTrigger);
+    }
+    const closeButton = document.querySelector('#close-button');
+    closeButton.addEventListener('click', this.eventToTrigger);
+  };
+  eventToTrigger = (event) => {
+    this.checkClickOnIcon(event);
+  };
+  checkClickOnIcon = (event) => {
+    const organismIdToClassMapping = {
```
Comment (Solnick) [inline src/Modal.js:45]:
```suggestion
    // comment here
    const organismIdToClassMapping = {
```
Please leave a comment here that the keys of this object are corresponding with the id of the animal element in the modal

File: src/Board.js (hunk 11 above)
Comment (nattsnb) [reply src/Board.js:83]: done

File: src/Modal.js (hunk 12 above)
Comment (nattsnb) [reply src/Modal.js:65]: done

File: src/Modal.js (hunk 13 above)
Comment (nattsnb) [reply src/Modal.js:45]: done
